### PR TITLE
Scheduler preemption logic as a PostFilter plugin

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What type of PR is this?**

/kind feature
/sig scheduling
/priority important-soon

**What this PR does / why we need it**:

- [x] Implement preemption logic as a PostFilter plugin
- [x] New unit tests and old tests have been migrated.

**Which issue(s) this PR fixes**:

Part of #91038.

**Special notes for your reviewer**:

- This PR doesn't enable the `defaultpreemption` plugin in the framework registry. Will remove the old hard-coded preemption logic and enable this plugin in a followup, as well as migrating current UTs and add new integration tests.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+		}
+		// TODO: accommodate eventing.
+		// prof.Recorder.Eventf(victim, preemptor, v1.EventTypeNormal, "Preempted", "Preempting", "Preempted by %v/%v on node %v", preemptor.Namespace, preemptor.Name, nodeName)
+	}
+
+	return framework.NewStatus(framework.Schedulable)
+}
+
+// CalStrategies calculates a slice of Strategy. Executing any Strategy can make the given <pod> schedulable.
+func CalStrategies(ctx context.Context, state *framework.CycleState, pod *v1.Pod,
```
Comment (damemi) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:147]: Is `Cal` short for `Call` (like `CallExtenders` below) or `Calc` (like calculate)?

Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:147]: It's short for "Calculate". Prefer "Calc"?

Comment (damemi) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:147]: ah, sure it's not a very big nit but I just wasn't sure if this was a typo

**Comment by ahg-g** [comment]
Thanks Wei, this is a large PR, I was hoping that we migrate the logic more gradually, and initially without introducing new features (I see the concept of strategy being introduced).

What I suggest we do to minimize the chances of regressions and make it easy to review is to create the PostFilter plugin as a wrapper around the preemption logic in the scheduler.go and generic_scheduler.go, even if we had to export all preemption-related functions, here is one approach:

1) To avoid cyclical dependency, we can move [preempt](https://github.com/kubernetes/kubernetes/blob/4cc44fbef626e87e59f7e9e7bd0bf9ec447a3c39/pkg/scheduler/scheduler.go#L388:25) and the few other functions it calls to generic_scheduler.go. scheduler.go would just invoke a single function from generic_scheduler.go that is synonym to RunPostFilterPlugins.

2) The default preempt plugin would initially just call this new ```Preempt``` function defined in generic_scheduler.go. 

3) We register the PostFilter and remove the call to ```Preempt``` in scheduler.go, and so preemption now runs through the PostFilter plugin, integration and e2e test should keep working.

4) We gradually move functions from generic_scheduler.go to the plugin package, each with its unit tests.

5) We refactor the old code to introduce new concepts if any (like strategy)

What do you think?

**Comment by Huang-Wei** [comment]
@ahg-g I also realized this PR a bit too large (labeled XXL)... I will split it into 2 or 3 PRs.

**Comment by Huang-Wei** [comment]
/hold
I will close this PR in favor of breaking it into following small PRs.

- [postfilter-impl-1] Refactor scheduler preempt interface #92009
- Expose EventRecorder to scheduler FrameworkHandle #92010
- [postfilter-impl-2] Introduce a defaultpreemption PostFilter plugin #92012
- [postfilter-impl-3] Activate defaultpreemption PostFilter plugin and disable hard-coded preemption call
- [postfilter-impl-4] Move `Preempt()` functions to defaultpreemption package.
- [postfilter-impl-5] Refactor `Preempt()` into unit-test-able functions.

**Comment by ahg-g** [comment]
Great plan, thanks.

**Pull request closed without merging**